File: packages/cli-kit/src/io.ts
import { CliError } from "./cli.ts";

/** Fixed-width text table from objects; columns default to the first row's keys. */
export function table(rows: Record<string, unknown>[], options: { columns?: string[]; max?: number; header?: boolean } = {}): string {
  if (!rows.length) return "(no results)";
  const columns = options.columns ?? Object.keys(rows[0]);
  const max = options.max ?? 60;
  const cell = (value: unknown) => { const text = value === null || value === undefined ? "" : String(value).replace(/\s+/g, " "); return text.length > max ? text.slice(0, max - 1) + "…" : text; };
  const widths = columns.map(column => Math.max(column.length, ...rows.map(row => cell(row[column]).length)));
  const line = (values: string[]) => values.map((value, i) => value.padEnd(widths[i])).join("  ").trimEnd();
  const lines = rows.map(row => line(columns.map(column => cell(row[column]))));
  if (options.header === false) return lines.join("\n");
  return [line(columns), line(widths.map(width => "-".repeat(width))), ...lines].join("\n");
}

export const json = (value: unknown): string => JSON.stringify(value, null, 2);

/** Lets the user pick one line with sk or fzf. Returns null when nothing was chosen or no picker exists. */
export function pick(lines: string[], options: { prompt?: string; multi?: boolean } = {}): string[] | null {
  if (!lines.length) return null;
  if (lines.length === 1) return [lines[0]];
  const bin = process.env.PF_PICKER ?? Bun.which("sk") ?? Bun.which("fzf");
  if (!bin) throw new CliError("no fuzzy picker: install sk or fzf, or pass an exact match");
  const args = [bin, ...(options.multi ? ["--multi"] : []), ...(options.prompt ? ["--prompt", options.prompt] : [])];
  const result = Bun.spawnSync({ cmd: args, stdin: Buffer.from(lines.join("\n") + "\n"), stdout: "pipe", stderr: "inherit" });
  const chosen = result.stdout.toString().split("\n").map(line => line.trim()).filter(Boolean);
  return chosen.length ? chosen : null;
}

/** Picks exactly one, resolving either an exact label or the fuzzy choice. */
export function pickOne<T>(items: T[], labelOf: (item: T) => string, options: { prompt?: string } = {}): T | null {
  if (!items.length) return null;
  if (items.length === 1) return items[0];
  const chosen = pick(items.map(labelOf), options);
  if (!chosen) return null;
  return items.find(item => labelOf(item) === chosen[0]) ?? null;
}

/** Opens a file, directory, or URL with the OS handler. */
export const openDefault = (target: string): number => Bun.spawnSync({ cmd: [process.platform === "darwin" ? "open" : "xdg-open", target], stdio: ["ignore", "inherit", "inherit"] }).exitCode ?? 1;

export const editor = (): string => process.env.EDITOR ?? Bun.which("nvim") ?? Bun.which("vim") ?? "vi";

/** Opens files in $EDITOR; vim-likes get vertical splits, at most `maxSplits`. */
export function editFiles(files: string[], maxSplits = 3): number {
  const bin = editor();
  const cmd = /vim/i.test(bin) ? [bin, files.length > maxSplits ? `-O${maxSplits}` : "-O", ...files] : [bin, ...files];
  return Bun.spawnSync({ cmd, stdio: ["inherit", "inherit", "inherit"] }).exitCode ?? 1;
}

export const copyToClipboard = (text: string): boolean => {
  const bin = process.env.PORTFOLIO_PBCOPY_BIN ?? Bun.which("pbcopy");
  if (!bin) return false;
  return (Bun.spawnSync({ cmd: [bin], stdin: Buffer.from(text) }).exitCode ?? 1) === 0;
};

export const readStdin = async (): Promise<string> => (await Bun.stdin.text()).replace(/\n$/, "");

/** y/N prompt; non-interactive runs answer no unless `assume` is set. */
export function confirm(question: string, assume?: boolean): boolean {
  if (assume !== undefined) return assume;
  if (!process.stdin.isTTY) return false;
  const answer = prompt(`${question} [y/N]`) ?? "";
  return /^y(es)?$/i.test(answer.trim());
}

export const numberArg = (value: string | undefined, what = "id"): number => {
  const n = Number(value);
  if (!value || !Number.isInteger(n) || n <= 0) throw new CliError(`${what} must be a positive integer, got ${JSON.stringify(value ?? "")}`);
  return n;
};

File: lib/pf/io.rb
module Pf
  class << self
    def json(value) = JSON.pretty_generate(value)

    def table(rows, columns: nil, max: 60, header: true)
      return '(no results)' if rows.empty?
      columns ||= rows.first.keys
      cell = lambda do |value|
        text = value.to_s.gsub(/\s+/, ' ')
        text.length > max ? text[0, max - 1] + '…' : text
      end
      values = rows.map { |row| columns.map { |column| cell.call(row[column]) } }
      widths = columns.each_index.map { |i| [columns[i].to_s.length, *values.map { |row| row[i].length }].max }
      line = ->(row) { row.each_with_index.map { |value, i| value.to_s.ljust(widths[i]) }.join('  ').rstrip }
      lines = values.map { |row| line.call(row) }
      lines.unshift(line.call(columns), line.call(widths.map { |width| '-' * width })) if header
      lines.join("\n")
    end

    def number_arg(value, what = 'id')
      number = Float(value, exception: false)
      unless number&.finite? && number.positive? && number == number.to_i
        raise Error, "#{what} must be a positive integer, got #{JSON.generate(value || '')}"
      end
      number.to_i
    end

    def which(name)
      ENV.fetch('PATH', '').split(File::PATH_SEPARATOR).map { |dir| File.join(dir, name) }.find { |path| File.file?(path) && File.executable?(path) }
    end

    def open_default(target)
      system(RUBY_PLATFORM.include?('darwin') ? 'open' : 'xdg-open', target)
      $?.exitstatus
    end

    def edit_files(paths, max_splits = 3)
      editor = ENV['EDITOR'] || which('nvim') || which('vim') || 'vi'
      args = editor.match?(/vim/i) ? [paths.length > max_splits ? "-O#{max_splits}" : '-O', *paths] : paths
      system(editor, *args)
      $?.exitstatus
    end

    def copy_to_clipboard(text)
      bin = ENV['PORTFOLIO_PBCOPY_BIN'] || which('pbcopy')
      return false unless bin
      _, status = Open3.capture2(bin, stdin_data: text)
      status.success?
    end

    def read_stdin = STDIN.read.delete_suffix("\n")

    def confirm(question, assume: nil)
      return assume unless assume.nil?
      return false unless STDIN.tty?
      print "#{question} [y/N]"
      STDIN.gets.to_s.strip.match?(/\Ay(?:es)?\z/i)
    end

    def pick_one(items, prompt: nil)
      return items.first if items.length <= 1
      bin = ENV['PF_PICKER'] || which('sk') || which('fzf')
      raise Error, 'no fuzzy picker: install sk or fzf, or pass an exact match' unless bin
      labels = items.map { |item| yield item }
      args = prompt ? ['--prompt', prompt] : []
      output, = Open3.capture2(bin, *args, stdin_data: labels.join("\n") + "\n")
      chosen = output.lines.map(&:strip).reject(&:empty?).first
      index = labels.index(chosen)
      index && items[index]
    end
  end
end
